clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/deploy/ops/CreateServerGroupAtomicOperation.java

```diff
@@ -54,88 +59,188 @@ private static Task getTask() {
+    AmazonECS ecs = getAmazonEcsClient();
+    String serviceName = getNextServiceName();
+    Collection<LoadBalancer> loadBalancers = new LinkedList<>();
+    loadBalancers.add(getLoadBalancer());
+
+    Integer desiredCapacity = getDesiredCapacity();
+    String taskDefinitionArn = taskDefinition.getTaskDefinitionArn();
+
+    DeploymentConfiguration deploymentConfiguration = new DeploymentConfiguration()
+      .withMinimumHealthyPercent(50)
+      .withMaximumPercent(100);
+
+    CreateServiceRequest request = new CreateServiceRequest()
+      .withServiceName(serviceName)
+      .withDesiredCount(desiredCapacity != null ? desiredCapacity : 1)
+      .withCluster(description.getEcsClusterName())
+      .withRole(description.getIamRole())
+      .withLoadBalancers(loadBalancers)
+      .withTaskDefinition(taskDefinitionArn)
+      .withDeploymentConfiguration(deploymentConfiguration);
+
+    updateTaskStatus(String.format("Creating %s of %s with %s for %s.",
+      desiredCapacity, serviceName, taskDefinitionArn, description.getCredentialAccount()));
+
+    Service service = ecs.createService(request).getService();
+
+    updateTaskStatus(String.format("Done creating %s of %s with %s for %s.",
+      desiredCapacity, serviceName, taskDefinitionArn, description.getCredentialAccount()));
+
+    return service;
+  }
+
+  private void createAutoScalingGroup(Service service) {
+    AWSApplicationAutoScaling autoScalingClient = getAmazonApplicationAutoScalingClient();
+    RegisterScalableTargetRequest request = new RegisterScalableTargetRequest()
+      .withServiceNamespace(ServiceNamespace.Ecs)
+      .withScalableDimension(ScalableDimension.EcsServiceDesiredCount)
+      .withResourceId(String.format("service/%s/%s", description.getEcsClusterName(), service.getServiceName()))
+      .withRoleARN(service.getRoleArn())
+      .withMinCapacity(0)
+      .withMaxCapacity(getDesiredCapacity());
+
+    updateTaskStatus("Creating Amazon Application Auto Scaling Scalable Target Definition...");
+    autoScalingClient.registerScalableTarget(request);
+    updateTaskStatus("Done creating Amazon Application Auto Scaling Scalable Target Definition...");
+  }
+
+  private DeploymentResult getDeploymentResult(Service service) {
+    Map<String, String> namesByRegion = new HashMap<>();
+    namesByRegion.put(getRegion(), service.getServiceName());
 
-    String region = description.getAvailabilityZones().keySet().iterator().next();
-    AmazonECS ecs = amazonClientProvider.getAmazonEcs(description.getCredentialAccount(), credentials.getCredentialsProvider(), region);
-    TaskDefinition taskDefinition = registerTaskDefinition(ecs, versionString);
+    DeploymentResult result = new DeploymentResult();
+    result.setServerGroupNames(Arrays.asList(getServerGroupName(service)));
+    result.setServerGroupNameByRegion(namesByRegion);
+    return result;
+  }
+
+  private LoadBalancer getLoadBalancer() {
```

Comment:
perhaps `retrieveLoadBalancer()` would be a better name.  Also, arguments to this method would make it easier to test

Reply:
Renamed. 

I do not want to add redundant arguments to this call. All of the information is contained in the object. This is not a static method so for testing you will need to initialize this with an appropriate description object. 

If you are worried about the `inferNextServerGroupVersion` the ContainerInformationService needs to be mocked anyways. We will likely to add additional constructors to be able to introduce Mock objects for testing.